### DISCUSSION TITLE
How to treat behavioral variables for CEBRA behavior

### DISCUSSION THREAD

**Discussion by GioLetti** [AdaptiveMotorControlLab/CEBRA#308] [Q&A]
Hi,

thanks for the super interesting and innovative paper. I have a quick question regarding the usage of CEBRA behavior. I'm using CEBRA to perform manifold learning on calcium imaging data during freely moving navigation, but I don't know exactly how CEBRA handles certain behavioral variables or how I should treat them. 

Specifically, I have the x-y mouse position on a 2D arena and the allocentric head direction in radians. Can I pass to the fit function as labels the data as a matrix [time_points, n_dims] with n_dims being x pos, y_pos, and head direction in radians? Or should I bin the variables and provide a linearized vector of positional bin number and head direction bin number? 

Thanks for your amazing work,

Giorgio

**Comment by stes** [comment]
Hi @GioLetti , as a quick first start I would recommend to use `x, y, cos direction, sin direction`. The reason is that the sampling assumes an approximately Gaussian distribution around the passed variables by default.

The more sophisticated option would be to implement a custom sampler that can handle angles. But I think for a first impression, the suggestion above is an acceptable pragmatic solution.

**Reply by stes** [reply]
Please see here
https://github.com/AdaptiveMotorControlLab/CEBRA/blob/d1842ccc659bdf2d458ab31784ae029b9d48d21f/cebra/distributions/continuous.py#L254-L298

and here

https://github.com/AdaptiveMotorControlLab/CEBRA/blob/d1842ccc659bdf2d458ab31784ae029b9d48d21f/cebra/distributions/continuous.py#L200-L238

**Reply by stes** [reply]
It could be an interesting contribution to the package to add a distribution that properly handles x,y and the head direction.

**Reply by GioLetti** [reply]
Hi @stes,

sorry for the late reply but I took some time to better understand the sampling process of CEBRA and make some tests.

Firstly, I've tried to use as auxiliary variable the HD encoded as sin and cos as you suggested and I got the following embeddings:

<img width="1336" height="477" alt="hd_embeddings" src="https://github.com/user-attachments/assets/b82dc545-0fdb-4b92-9048-f74928ea86d0" />

If I use all the data of one recording I'm able to get a nearly perfect ring, while if I divided in train and test the embeddings are still resembling a circle, but more spread. Do you have any idea of a possible explanation?

Moreover, I've tried to use again the HD values but now binned circularly, and I still get a near-perfect ring using all the data. I'm know wondering if these results are due to the autocorrelation of the behavior and calcium traces, as I've used a time_offset=10, so the positive samples are always very close.  

Moreover, as I want to study position and HD that in freely moving animal are correlated and dependent I've tried different approaches: 
1. I used as auxiliary variable x,y,sin, and cos values 
2. x,y as continuous values while HD binned and passed as categorical value.

For what I understood by looking at the supplementary of the paper, the 2 approach should induce the model to treat position and HD as dependent during positive sample sampling, am I correct? 

For all these tests I always used the model with offset 10 and time_offset 10 and temperature fixed at 1. 

Coming back to the possibility of implementing sampler specific for HD, If I correctly understood I should modify just the TimeDeltaDistribution class ( at least for my specific case). But I'm still not sure if the sampling is done using the vector of indeces, as if this the case actually I would not need to take care of the problem with the angular values, so I'm a bit stuck on this.

Sorry for the long reply, and thank you again for your kind answer.

Giorgio

**Reply by GioLetti** [reply]
Hi @stes,

I've implemented a TimedeltaDistribution variant for circular variable for single-session only. Head direction is encoded as you suggested as cos and sin. I'm using rotation matrix instead of additive deltas when computing time_difference, so that the queries remains on the unit circle and the 0°/360° continuity is handled. It seems to work in isolated tests. I'd like your input on the right way to extend this to a mixed circular + linear setting (e.g. [x, y, cos θ, sin θ]).

I have a few other questions regarding CEBRA:

1. DistanceMatrix/ContinuousIndex compute plain squared Euclidean distance over however many columns the continuous index has, with no per-dimension normalization. So if I pass [x, y, cos θ, sin θ] with x,y ∈ [0, 44] cm and cos θ, sin θ ∈ [-1, 1], position dominates the distance and positive-pair selection mostly reflects position, not head direction. Should I normalize each dimension to unit variance before combining into the auxiliary matrix passed to CEBRA model fit function? Is there internal handling of this I might be missing, or is this left entirely to the user?

2. For strongly autocorrelated modalities like calcium imaging during freely moving navigation both the behavior and the delta F/F traces are autocorrelated. Did you validate CEBRA-Behavior's embedding/decoding quality against a null that preserves the label's own autocorrelation but breaks the true correspondence to neural activity (circular shift of behavior), rather than only a full i.i.d. shuffle? I'm trying to understand how much the autocorrelation can impact the results and eventually provide good ones even if genuine tuning is not present.

3. Was time_offset selection in the paper/demos informed by the autocorrelation timescale of the neural signal (e.g. calcium indicator decay) as well as the behavior variable? I consistently see embedding quality degrade as time_offset grows, with maximum degradation close to my calcium signal's autocorrelation decay time. Did you run a time_offset sweep considering signal autocorrelation?

4. I'm performing a cross-validation procedure for fitting the embeddings on the same recording to evaluate across fold consistency. To do so I'm using sklearn KFold class with shuffle parameter set to True. I saw in another discussion that you were suggesting avoiding such approach particularly with the model architecture 'offset10-model' as the sampler needs time contigous data. At the same time even If I use shuffle= True, most of the training data are contiguous with just few points being at a maximum of 5 timesteps apart. Do you think this minimal distance can still be problematic for the model? Do you have any recommendations on how to perform such consistency analysis?

**Reply by stes** [reply]
Hi @GioLetti, thanks for following up and the extensive questions. I will follow up on this with some suggestions in the next days.

**Reply by GioLetti** [reply]
Hi @stes, thank you for the kind reply. I would like to start a more in-depth analysis and I would like to set it up as correctly as possible. Really appreciate you taking the time for those questions.
